src/main/java/by/dragonsurvivalteam/dragonsurvival/client/handlers/ClientFlightHandler.java

```diff
@@ -247,241 +252,233 @@ private static void spawnSpinParticle(Player player, ParticleOptions particleDat
+				*/
+				levitationLeft = Functions.secondsToTicks(levitationAfterEffect);
+			} else if (levitationLeft > 0) {
+				// TODO :: Set to 0 once ground is reached?
+				if (event.phase == Phase.END) {
+					levitationLeft--;
+				}
+			} else {
+				DragonStateProvider.getCap(player).ifPresent(handler -> {
+					if (handler.isDragon()) {
+						Vec3 viewVector = player.getLookAngle();
+						double yaw = Math.toRadians(player.getYHeadRot() + 90);
+
+						// Only apply while in water (not while flying)
+						if (ServerFlightHandler.canSwimSpin(player) && ServerFlightHandler.isSpin(player)) {
+							Input movement = player.input;
+
+							Vec3 deltaMovement = player.getDeltaMovement();
+
+							double maxFlightSpeed = ServerFlightHandler.maxFlightSpeed;
+							// FIXME :: Magic numbers at various places
+							ax = Mth.clamp(ax, -0.2 * maxFlightSpeed, 0.2 * maxFlightSpeed);
+							az = Mth.clamp(az, -0.2 * maxFlightSpeed, 0.2 * maxFlightSpeed);
+
+							// Increase acceleration depending on how sharply the player turns their character
+							ax += Math.cos(yaw) / 500 * 50;
+							az += Math.sin(yaw) / 500 * 50;
+							ay = viewVector.y / 8;
+
+							if (viewVector.y < 0) {
+								deltaMovement = deltaMovement.add(ax, 0, az);
+							} else {
+								// Only increase height if the player is looking up
+								deltaMovement = deltaMovement.add(ax, ay, az);
+							}
 
-						if(lookY < 0){
-							motion = motion.add(ax, 0, az);
-						}else{
-							motion = motion.add(ax, ay, az);
-						}
-						motion = motion.multiply(0.99F, 0.98F, 0.99F);
+							// TODO :: Why 0.99 etc.?
+							deltaMovement = deltaMovement.multiply(0.99F, 0.98F, 0.99F);
 
-						if(motion.length() != player.getDeltaMovement().length()){
-							NetworkHandler.CHANNEL.sendToServer(new SyncFlightSpeed(player.getId(), motion));
+							player.setDeltaMovement(deltaMovement);
+							ay = player.getDeltaMovement().y;
 						}
 
-						player.setDeltaMovement(motion);
-						ay = player.getDeltaMovement().y;
-					}
-
-					if(dragonStateHandler.isWingsSpread()){
-						Input movement = player.input;
-						boolean hasFood = player.getFoodData().getFoodLevel() > ServerFlightHandler.flightHungerThreshold || player.isCreative() || ServerFlightHandler.allowFlyingWithoutHunger;
-
-
-						if(!hasFood){
-							ay = Mth.clamp(Math.abs(ay * 4), -0.2 * ServerFlightHandler.maxFlightSpeed, 0.2 * ServerFlightHandler.maxFlightSpeed);
-						}
+						if (handler.isWingsSpread()) {
+							Input movement = player.input;
+							boolean hasFood = player.getFoodData().getFoodLevel() > ServerFlightHandler.flightHungerThreshold || player.isCreative() || ServerFlightHandler.allowFlyingWithoutHunger;
 
-						//start
-						if(ServerFlightHandler.isFlying(player)){
-							if(!wasFlying){
-								wasFlying = true;
+							if (!hasFood) {
+								// TODO :: If you use Math.abs you always get a positive number, shouldn't this be max() instead of clamp()?
```

Comment:
many crutches have been made to make the inedible edible. Maybe this is one of them